homeassistant/components/switch/fritzbox_callforwarding.py

```diff
@@ -0,0 +1,180 @@
+
+
+class CallForwarding:
+    """Representation of call forwarding entity."""
+
+    def __init__(self, fritz_box, uid, from_number,
+                 to_number, connection_type, enabled):
+        """Initialize the call forwarding entity."""
+        self.fritz_box = fritz_box
+        self.name = "callforwarding_" + uid
+        self.uid = uid
+        self.from_number = from_number
+        self.to_number = to_number
+        self.connection_type = connection_type
+        self.enabled = enabled
+
+    def enable(self):
+        """Enable the call forwarding for a uid."""
+        kwargs = {'NewDeflectionId': self.uid, 'NewEnable': 1}
+        self.fritz_box.call_action(
+            'X_AVM-DE_OnTel:1', 'SetDeflectionEnable', **kwargs)
+        self.enabled = 1
+
+    def disable(self):
+        """Disable the call forwarding for a uid."""
+        kwargs = {'NewDeflectionId': self.uid, 'NewEnable': 0}
+
+        self.fritz_box.call_action(
+            'X_AVM-DE_OnTel:1', 'SetDeflectionEnable', **kwargs)
+        self.enabled = 0
+
+    def __str__(self):
+        """Create a string representation of a CallForwarding."""
+        return "%s --> %s: %s" % (self.from_number,
+                                  self.to_number, self.enabled)
+
+
+class FritzCallForwardingSwitch(SwitchDevice):
+    """Representation of a FRITZ!CallForwarding switch."""
+
+    def __init__(self, call_forwarding):
+        """Initialize the switch with a call forwarding from the fritzbox."""
+        self._call_forwarding = call_forwarding
+
+    @property
+    def name(self):
+        """Return the name of the FRITZ!CallForwarding switch, if any."""
+        return self._call_forwarding.name
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return ICON
+
+    @property
+    def is_on(self):
+        """Return true if switch is on."""
+        return bool(self._call_forwarding.enabled)
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        self._call_forwarding.enable()
+
+    def turn_off(self):
+        """Turn the switch off."""
+        self._call_forwarding.disable()
+
+    @Throttle(MIN_TIME_BETWEEN_SCANS)
+    def update(self):
+        """Get the latest call forwarding data
+        from the fritz box and updates the states."""
+        self._call_forwarding = get_call_forwarding_by_uid(
+            self._call_forwarding.fritz_box, self._call_forwarding.uid)
+        _LOGGER.debug(self._call_forwarding)
+
+    def __str__(self):
+        """Create a string representation of a FritzCallForwardingSwitch."""
+        return str(self._call_forwarding)
+
+
+def get_call_forwardings(fritz_box):
+    """Get a list with all CallForwarding Objects ."""
+    import xml.etree.ElementTree as ET
+
+    call_forwardings = []
+    deflections = fritz_box.call_action('X_AVM-DE_OnTel:1', 'GetDeflections')
+    deflection_list = ET.fromstring(deflections['NewDeflectionList'])
```

Comment:
This is definitely too much protocol specific info. This needs to be moved to a Python package, either FritzConnection or another one.